File: csharp/demo/demo23a02-thread-local.cs
/*
 * THREAD-LOCAL STORAGE
 * Introduction
 *
 * Use valueFactory function for better initialization
 */
using System;
using System.Threading;



class Demo22302 : IRunnable
{
    public void run()
    {
        // Main thread sets value = "APPLE"
        MyTask.set("APPLE");
        Console.WriteLine(MyTask.get());

        // Child thread gets value
        // Expected output: "NOT SET"
        new Thread(() =>
        {
            Console.WriteLine(MyTask.get());
        }).Start();
    }


    class MyTask
    {
        private static ThreadLocal<string> data = new ThreadLocal<string>(() => "NOT SET");

        public static string get()
        {
            return data.Value;
        }

        public static void set(string value)
        {
            data.Value = value;
        }
    }
}

File: cpp/cpp-std/demo23a02-thread-local.cpp
/*
THREAD-LOCAL STORAGE
Introduction
    The smart-pointer way to use thread-local storage
*/


#include <iostream>
#include <string>
#include <memory>
#include <thread>
using namespace std;



thread_local std::shared_ptr<string> value;



string getValue() {
    if (nullptr == value.get()) {
        value.reset(new string("NOT SET"));
    }

    return *value.get();
}



void doTask() {
    cout << getValue() << endl;
}



int main() {
    // Main thread sets value = "APPLE"
    value.reset(new string("APPLE"));
    cout << getValue() << endl;

    // Child thread gets value
    // Expected output: "NOT SET"
    auto th = std::thread(&doTask);
    th.join();

    return 0;
}
